api/src/main/java/org/openmrs/module/fgh/mpi/MpiContext.java

```diff
@@ -0,0 +1,246 @@
+package org.openmrs.module.fgh.mpi;
+
+import org.apache.commons.lang3.StringUtils;
+import org.openmrs.api.APIException;
+import org.openmrs.api.AdministrationService;
+import org.openmrs.api.context.Context;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.security.*;
+import java.security.cert.CertificateException;
+
+import static org.openmrs.module.fgh.mpi.MpiConstants.*;
+
+public class MpiContext {
+	
+	private static final Logger log = LoggerFactory.getLogger(MpiContext.class);
+	
+	private TokenInfo tokenInfo;
+	
+	private String contentType;
+	
+	private AuthenticationType authenticationType;
+	
+	private boolean contextInitialized;
+	
+	private SSLContext sslContext;
+	
+	private String serverBaseUrl;
+	
+	private MpiSystemType mpiSystem;
+	
+	private String openmrsUuidSystem;
+	
+	private String clientId;
+	
+	private String clientSecret;
+	
+	public static MpiContext mpiContext;
+	
+	public MpiContext() {
+	}
+	
+	public static synchronized MpiContext initIfNecessary() throws Exception {
+		if (mpiContext == null) {
+			mpiContext = new MpiContext();
+			
+			mpiContext.init();
+		}
+		
+		return mpiContext;
+	}
+	
+	protected void init() throws Exception {
+		
+		synchronized (this) {
```

Comment:
By the way you changed the logic here, you are calling this method from a synchronized method, it means you no longer need this synchronized bock